src/test/scala/tech/cryptonomic/conseil/tezos/TezosDatabaseBigMapTest.scala

```diff
@@ -0,0 +1,225 @@
+package tech.cryptonomic.conseil.tezos
+
+import scala.concurrent.duration._
+import org.scalatest.Matchers
+import com.typesafe.scalalogging.LazyLogging
+import org.scalatest.concurrent.IntegrationPatience
+import org.scalatest.WordSpec
+import org.scalatest.concurrent.ScalaFutures
+import slick.jdbc.PostgresProfile.api._
+import tech.cryptonomic.conseil.util.RandomSeed
+import tech.cryptonomic.conseil.tezos.TezosTypes.{
+  Contract,
+  Decimal,
+  Micheline,
+  Operation,
+  OperationsGroup,
+  Origination,
+  ScriptId,
+  Transaction
+}
+import tech.cryptonomic.conseil.tezos.Tables.BigMapContentsRow
+import tech.cryptonomic.conseil.tezos.Tables.BigMapsRow
+import com.softwaremill.diffx.scalatest.DiffMatcher._
+import tech.cryptonomic.conseil.tezos.Tables.OriginatedAccountMapsRow
+
+class TezosDatabaseBigMapTest
+    extends WordSpec
+    with TezosDataGeneration
+    with InMemoryDatabase
+    with Matchers
+    with ScalaFutures
+    with LazyLogging
+    with IntegrationPatience {
+
+  "The database api" should {
+      //simplify signatures
+      type ListTransf[A] = List[A] => List[A]
+      //needed for most tezos-db operations
+      import scala.concurrent.ExecutionContext.Implicits.global
+
+      val sut = TezosDatabaseOperations
+
+      "save big map diffs allocations and updates contained in a list of blocks" in {
```

Comment:
Could you please explain what does `diff` in this context mean? I cannot understand it by reading this test. Second thing is that those two tests are pretty long hence they're hard to read. Unfortunately, I don't understand them enough to give an idea of how to split them.

Reply:
Yes, sure, let me explain a bit.

Each operation result (internal or external), might modify the values within a big_map, which is sort of a custom type for lazily evaluated key-values that would require much memory. Each contract can have its own data defined there.

What an operation does is to modify such map, by defining a new one, removing it, changing the content, copying the content to another contract's big_map.
That's why you only have diffs included in the json, I guess to save space, instead of resending the whole map on every operation. The big_map_diff stores the changes that happened to the big_map, and we apply those to our stored data.

The tests currently verify the possible diffs pair-wise, to save space.
The first checks what happens if there was map "allocation" and an "update".
The second is the case of map "copy" and "remove".

The tests are complex because we need to generate blocks which contains sample operations that have the appropriate big_map_diff values inside, and I injected those inside the pre-generated generic blocks used in database testing.
Also, for the second case, we need data already stored in the db before being able to apply the changes.